### PR TITLE
[inductor] Iterative percolate tags

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117306

Fixes https://github.com/pytorch/pytorch/issues/116581

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -814,10 +814,15 @@ def run_node(self, node) -> Any:
             last_node = min(indices, key=lambda tup: tup[0])[1]
 
         def percolate_tags(node, recompute_tag):
-            for arg in node.all_input_nodes:
-                if hasattr(arg, "meta"):
+            queue = [node]
+            visited = set()
```
Comment (aorenste) [inline torch/_inductor/pattern_matcher.py:818]: nit: Should probably init `visited` with `node`:
```
visited = set([node])
```

**Review by aorenste — APPROVED** [review]
Behavior is slightly different (in a graph this won't visit nodes multiple times whereas the original would) but I suspect this is better.

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (aorenste) [reply torch/_inductor/pattern_matcher.py:818]: Nevermind - you're adding to visited after popping.

**Comment by anijain2305** [comment]
> Behavior is slightly different (in a graph this won't visit nodes multiple times whereas the original would) but I suspect this is better.

Yes, we don't even want to visit multiple times (I suspect that could be the reason Stas saw deep recursive stacks). Just visiting once is enough.

File: torch/_inductor/pattern_matcher.py [hunk 2]
```diff
@@ -814,10 +814,15 @@ def run_node(self, node) -> Any:
...
+
+            while queue:
+                arg = queue.pop()
+                if arg not in visited and hasattr(arg, "meta"):
+                    visited.add(arg)
                     arg.meta["recompute"] = recompute_tag
-                    percolate_tags(arg, recompute_tag)
+                    queue.extend(arg.all_input_nodes)
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:825]: TIL about `all_input_nodes`

File: torch/_inductor/pattern_matcher.py [hunk 3]
```diff
@@ -814,10 +814,15 @@ def run_node(self, node) -> Any:
             last_node = min(indices, key=lambda tup: tup[0])[1]
 
         def percolate_tags(node, recompute_tag):
-            for arg in node.all_input_nodes:
-                if hasattr(arg, "meta"):
+            queue = [node]
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:817]: previously, the first `node` would not have `arg.meta["recompute"] = recompute_tag` set. was that intentional?

**Review by eellison — APPROVED** [review]

File: torch/_inductor/pattern_matcher.py (hunk 3 above)
Comment (anijain2305) [reply torch/_inductor/pattern_matcher.py:817]: Yes, that was.

**Pull request closed without merging**